ethstorage/downloader/blob_disk_cache.go

```diff
@@ -0,0 +1,255 @@
+		return err
+	}
+	c.lookup[block.hash] = id
+	for _, b := range block.blobs {
+		c.index[b.kvIndex.Uint64()] = id
+		c.lg.Debug("Indexing blob in cache", "kvIdx", b.kvIndex, "hash", b.hash, "id", id)
+	}
+	c.lg.Info("Set blockBlobs to cache", "block", block.number, "id", id)
+	return nil
+}
+
+func (c *BlobDiskCache) Blobs(hash common.Hash) []blob {
+	c.mu.RLock()
+	id, ok := c.lookup[hash]
+	if !ok {
+		c.mu.RUnlock()
+		return nil
+	}
+	block, err := c.getBlockBlobsById(id)
+	c.mu.RUnlock()
+	if err != nil || block == nil {
+		return nil
+	}
+	c.lg.Info("Blobs from cache", "block", block.number, "id", id)
+	res := []blob{}
+	for _, blob := range block.blobs {
+		res = append(res, *blob)
+	}
+	return res
+}
+
+func (c *BlobDiskCache) GetKeyValueByIndex(idx uint64, hash common.Hash) []byte {
+	blob := c.getBlobByIndex(idx)
+	if blob != nil &&
+		bytes.Equal(blob.hash[0:ethstorage.HashSizeInContract], hash[0:ethstorage.HashSizeInContract]) {
+		return blob.data
+	}
+	return nil
+}
+
+// Access without verification through a hash: only for miner sampling
+func (c *BlobDiskCache) GetKeyValueByIndexUnchecked(idx uint64) []byte {
+	blob := c.getBlobByIndex(idx)
+	if blob != nil {
+		return blob.data
+	}
+	return nil
+}
+
+func (c *BlobDiskCache) getBlobByIndex(idx uint64) *blob {
+	c.mu.RLock()
+	id, ok := c.index[idx]
+	if !ok {
+		c.mu.RUnlock()
+		return nil
+	}
+	block, err := c.getBlockBlobsById(id)
+	c.mu.RUnlock()
+	if err != nil || block == nil {
+		return nil
+	}
+	for _, blob := range block.blobs {
+		if blob != nil && blob.kvIndex.Uint64() == idx {
+			return blob
+		}
+	}
+	return nil
+}
+
+func (c *BlobDiskCache) Cleanup(finalized uint64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for hash, id := range c.lookup {
+		block, err := c.getBlockBlobsById(id)
+		if err != nil {
+			c.lg.Error("Failed to get block from id", "id", id, "err", err)
+			continue
+		}
+		if block != nil && block.number <= finalized {
+			if err := c.store.Delete(id); err != nil {
+				c.lg.Error("Failed to delete block from id", "id", id, "err", err)
+			}
+			delete(c.lookup, hash)
+			for _, blob := range block.blobs {
+				if blob != nil && blob.kvIndex != nil {
+					delete(c.index, blob.kvIndex.Uint64())
+				}
+			}
+			c.lg.Info("Cleanup deleted", "finalized", finalized, "block", block.number, "id", id)
+		}
+	}
+}
+
+func (c *BlobDiskCache) getBlockBlobsById(id uint64) (*blockBlobs, error) {
+	data, err := c.store.Get(id)
+	if err != nil {
+		c.lg.Error("Failed to get block from id", "id", id, "err", err)
+		return nil, err
+	}
+	if len(data) == 0 {
+		c.lg.Warn("BlockBlobs not found", "id", id)
+		return nil, fmt.Errorf("not found: id=%d", id)
+	}
+	item := new(blockBlobs)
+	if err := rlp.DecodeBytes(data, item); err != nil {
+		c.lg.Error("Failed to decode block", "id", id, "err", err)
+		return nil, err
+	}
+	return item, nil
+}
+
+func (c *BlobDiskCache) Close() error {
```

Comment:
When we close the cache, we may want to clean the cache anyway, and the downloader will download the blobs from last saved finalized block when restart